[kjw142857] iP

kjw142857 opened the pull request:
# Sleepy

What is Sleepy? Sleepy simulates a student who is suffering from lack of sleep ~~due to the CS2103T iP~~. 

You should use Sleepy if:

- You are sleepy
- You are so sleepy that you **need** an AI chatbot to help you
- Did I mention that you must be sleepy?

What you need to do to set up Sleepy:
1. Sleep
2. Sleep some more
3. Install [Gradle](https://gradle.org/install/)
4. `git clone` this repository to your local computer
5. Navigate to the repository in your Command-Line-Interface (CLI)
6. Enter the command `gradlew build run`
7. Add and remove deadlines, events and todos!
8. Sleep

For instance:

- [x] Complete this pull request description
- [ ] Withdraw from CS2103T (by: Recess Week)
- [ ] Drop out of NUS
- [ ] Get some sleep (from: 5am to: 6am)

> I just want to sleep. - The author of this pull request

You can also improve the Java code in Sleepy if you feel that it can be even better! Here is an example:
``` java
public class Person {
    private boolean socialLife;
    private int sleepHours;
    public Person(boolean isSoftwareEngineer) {
       if (isSoftwareEngineer) {
           socialLife = false;
           sleepHours = 0;
       } else {
           socialLife = true;
           sleepHours = 8;
       }
    }
}
```

Or _alternatively,_ you can choose to sleep... 😪 😪 😪

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

kjw142857 committed:
Level-0

kjw142857 committed:
Revert "Level-0"

This reverts commit d60e29ba0ca7cd5aa2c6703a111e8b4b44524239.

kjw142857 committed:
Update Duke.java

kjw142857 committed:
Level-1

kjw142857 committed:
Level-2

kjw142857 committed:
Level-3

kjw142857 committed:
Level-4

kjw142857 committed:
Level-5

kjw142857 committed:
Level-6

kjw142857 committed:
A-TextUiTesting

kjw142857 committed:
Level-0

kjw142857 committed:
Add packages and header comments for classes

kjw142857 committed:
Update Sleepy.java

kjw142857 committed:
Implement saving of data in hard disk

kjw142857 committed:
Implement formatting of dates

kjw142857 committed:
A-MoreOOP

kjw142857 committed:
Change package imports

kjw142857 committed:
Merge branch 'add-gradle-support'

kjw142857 committed:
A-Gradle

kjw142857 committed:
A-JUnit

kjw142857 committed:
A-Jar

kjw142857 committed:
Level-9

kjw142857 committed:
A-JavaDoc

kjw142857 committed:
A-JavaDoc

kjw142857 committed:
Merge branch 'branch-A-JavaDoc'

kjw142857 committed:
A-CodingStandard

kjw142857 committed:
Merge branch 'branch-A-CodingStandard'

ThitipatC commented on src/main/java/sleepy/tasks/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @author kjw142857
+ */
+public class Deadline extends Task {
```
Would renaming to DeadlineTask be better? To me Deadline alone does not give the sense of it being a task.

kjw142857 replied:
Hmm, I think you make a very good point here, will consider your suggestion!

ThitipatC commented on src/main/java/sleepy/taskstorage/Storage.java:
```diff
@@ -0,0 +1,65 @@
+     * @return Tasks read from the file.
+     */
+    public ArrayList<String> readFile() {
+        ArrayList<String> readTasks = new ArrayList<>();
```
Not sure if it was meant to be read(as in participle or present)Tasks. Personally, tasksRead might be clearer.

kjw142857 replied:
I see! I would personally interpret `tasksRead` as a verb :sweat_smile: but I can totally see how it would mean the present participle, thanks for the input!

ThitipatC commented on src/main/java/sleepy/tasks/Event.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @author kjw142857
+ */
+public class Event extends Task {
```
Would renaming to EventTask be better? Similar reason to the Deadline class comment

ThitipatC commented on src/main/java/sleepy/tasks/ToDo.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @author kjw142857
+ */
+public class ToDo extends Task {
```
Would renaming to ToDoTask be better? Similar reason to the previous comments but in this case I felt like ToDo is more of a collection class (although without s) than a task.

kjw142857 committed:
Add checkstyle and fix checkstyle violations

kjw142857 committed:
Implement GUI for Sleepy Chatbot

kjw142857 committed:
Fix serious dependency issues

The java version in this project was erroneously set to 17, even though it was labelled "11". As such, I have:
- Changed the java version to SDK 11
- Remodelled some case-switch statements which would have cause a compilation error in Java 11, but did not in Java 17
- Created a new .jar file and fixed an issue with creation of directories for the hard disk file

kjw142857 committed:
Add assertions to iP code

Add assertions to iP code

The iP code needs assertions to check at key points of the code if the integral elements, such as the parser and chatbot cache, have failed.

These assertions will enable verification that the components of the Sleepy chatbot are running as intended.

kjw142857 committed:
Improve code quality and fix bugs

This commit aims to improve overall code readability, as well as fix some unexpected errors in the UI.

Some highlights:
- Significantly shortened some methods to follow the 30 LoC rule (excluding comments).
- Removed confusing abstractions, such that all things relating to string checking are handled by the parser.
- Improve labelling of variables.
- Fix a bug where upon reloading save from the hard disk, all tasks would be labelled as undone.

kjw142857 committed:
Merge pull request #2 from kjw142857/branch-A-CodeQuality

Improve code quality and fix bugs

kjw142857 committed:
Merge branch 'master' into branch-A-Assertions

kjw142857 committed:
Update TaskList.java

kjw142857 committed:
Merge pull request #1 from kjw142857/branch-A-Assertions

Add assertions to iP code

kjw142857 committed:
Update Parser.java

kjw142857 committed:
Add DoAfterTasks (Plans)

Add plans to the chatbot, where a user can indicate that they want to do a task after an event or date.

Ty-stan0417 commented on src/main/java/sleepy/Sleepy.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Constructor for the Sleepy class.
+     */
+    public Sleepy() {
+        String filePath = "./storage/HardDiskStorage.txt";
+        // Retrieve saved data
+        taskList = new TaskList(filePath);
+        isInExitState = false;
+    }
+
```
Consider to [specify file paths in an OS-independent way](https://www.sghill.net/how-do-i-make-cross-platform-file-paths-in-java.html).

kjw142857 replied:
I never knew about that, thank you for enlightening me on this! Will seriously consider implementing this to support different OSes.

Ty-stan0417 commented on src/main/java/sleepy/tasks/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+     * @param description Description of the deadline details.
+     * @param givenDeadline When the given deadline is.
+     */
+    public Deadline(String description, String givenDeadline) {
+        super(description);
+        this.givenDeadline = givenDeadline;
+        try {
+            formattedDate = LocalDate.parse(givenDeadline);
+        } catch (DateTimeParseException d) {
+            formattedDate = null;
+        }
+    }
```
LGTM. I like the way you construct your constructor and parse the String to LocalDate inside the constructor.

Ty-stan0417 commented on src/main/java/sleepy/tasks/Event.java:
```diff
@@ -0,0 +1,67 @@
+     * @return Description of this event.
+     */
+    @Override
+    public String getDescription() {
+        String startDate = formattedFromDate == null
+                ? start : formattedFromDate.format(DateTimeFormatter.ofPattern("d MMM yyyy"));
+        String endDate = formattedToDate == null
+                ? end : formattedToDate.format(DateTimeFormatter.ofPattern("d MMM yyyy"));
+        return "[E]" + super.getDescription() + " (from: " + startDate + " to: " + endDate + ")";
+    }
```
As you use `DateTimeFormatter.ofPattern("d MMM yyyy")` several time, you may consider to set this as a class constant to simplify your code.

kjw142857 replied:
That is a good suggestion, will look into doing it!

Ty-stan0417 commented on src/main/java/sleepy/taskstorage/Storage.java:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @param filePath filePath of the file used for storage.
+     */
+    public Storage(String filePath) {
+        try {
+            hardDiskFile = new File(filePath);
+            if (!hardDiskFile.exists()) {
+                Files.createDirectories(hardDiskFile.toPath().getParent());
+                hardDiskFile.createNewFile();
+            }
+        } catch (IOException i) {
+            ResponseHandler.appendLineToResponse(
+                    "Sleepy encountered a serious error! Please restart the application :("
+            );
+        }
+    }
```
Good handling about the file not exists and directory not exists.

Ty-stan0417 commented on src/main/java/sleepy/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,227 @@
+     * @param parsedCommand Access command given.
+     * @throws IllegalArgumentException If the command format is invalid.
+     */
+    public void access(String[] parsedCommand) throws IllegalArgumentException {
+        String command = parsedCommand[0];
+        switch (command) {
+        case "list":
+            printTasks();
+            break;
+        case "mark":
+            // Fallthrough
+        case "unmark":
+            // Fallthrough
+        case "delete":
+            int taskNumber = Integer.parseInt(parsedCommand[1]);
+            if (taskNumber < 1 || taskNumber > tasks.size()) {
+                throw new IllegalArgumentException("Invalid task number!");
+            }
+            handleTask(command, taskNumber);
+            break;
+        case "find":
+            String keywords = parsedCommand[1];
+            findTask(keywords);
+            break;
+        case "todo":
+            // Fallthrough
+        case "deadline":
+            // Fallthrough
+        case "plan":
+            // Fallthrough
+        case "event":
+            addTask(parsedCommand);
+            break;
+        default:
+            // Should never reach here - parser should have caught invalid command
+            assert false : "Parser did not parse command correctly";
+            throw new IllegalArgumentException("Incorrect command!");
+        }
+    }
```
LGTM. Your code looks very neat and easy to read.

Ty-stan0417 commented on src/main/java/sleepy/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,227 @@
+     * @param taskNumber Task number to be deleted.
+     * @throws IllegalArgumentException If the task number is invalid.
+     */
+    public void deleteTask(int taskNumber) throws IllegalArgumentException {
+        if (taskNumber <= 0 || taskNumber > tasks.size()) {
+            throw new IllegalArgumentException("Invalid task number!");
+        }
+        Task removedTask = tasks.remove(taskNumber - 1);
+        storage.saveTasks(tasks);
+        ResponseHandler.appendLineToResponse("Noted. I've removed this task:");
+        ResponseHandler.appendLineToResponse("  " + removedTask.getDescription());
+        ResponseHandler.appendLineToResponse(String.format("Now you have %d task(s) in the list.", tasks.size()));
+    }
+
+    /**
+     * Marks a specific task in this list as done, and prints its status.
+     *
+     * @param taskNumber Number (in this list) of the task to be marked.
+     */
+    public void markTaskAsDone(int taskNumber) {
+        Task targetTask = tasks.get(taskNumber - 1);
+        targetTask.markAsDone();
+        storage.saveTasks(tasks);
+        if (!isStartingUp) {
+            ResponseHandler.appendLineToResponse("Nice! I've marked this task as done:");
+            ResponseHandler.appendLineToResponse(targetTask.getDescription());
+        }
+        assert targetTask.isDone() == "true" : "Task was not marked as done correctly!";
+    }
```
LGTM. Good naming. I can know what is your method doing through your naming.

Ty-stan0417 reviewed (commented):
Overall, Good Job!👌

leeweiming3 commented on src/main/java/sleepy/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * Creates and adds a task to this list.
+     *
+     * @param parsedTask Task to be added, in the form of a string.
+     * @throws IllegalArgumentException If the string does not contain a task type.
+     */
+    public void addTask(String[] parsedTask) throws IllegalArgumentException {
+        String taskType = parsedTask[0];
```
A bit vague, as the variable is not a String but a collection of strings so should be plural

kjw142857 replied:
Thank you for pointing this out, will fix it ASAP!

leeweiming3 reviewed (commented):
LGTM. Very few code style errors

leeweiming3 commented on src/main/java/sleepy/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,227 @@
+     */
+    public void printTasks() {
+        if (tasks.size() == 0) {
+            ResponseHandler.appendLineToResponse("Your task list is empty! Looks like you can go back to sleep.");
```
Keep line length less than 110 chars

kjw142857 replied:
Thanks! The checkstyle was only catching >120 chars violations.

kjw142857 committed:
Multiple bug fixes and code quality improvements

kjw142857 committed:
Update README.md

kjw142857 committed:
Update HardDiskStorage.txt

kjw142857 committed:
Update .gitignore

kjw142857 committed:
Delete HardDiskStorage.txt

kjw142857 committed:
Improve Text GUI display for long lists

kjw142857 committed:
Add support for resizing GUI window

kjw142857 committed:
Add check for order of start/end dates

kjw142857 committed:
Add more unit tests

kjw142857 committed:
Fix minor typos in code

kjw142857 committed:
Fix bug for null check in formatted event dates

kjw142857 committed:
Add error handling for empty task description

kjw142857 committed:
Add User Guide

kjw142857 committed:
Update empty result for find command

kjw142857 committed:
Minor GUI updates

kjw142857 committed:
Update User Guide

kjw142857 committed:
Update User Guide

kjw142857 committed:
Update User Guide

kjw142857 committed:
Update User Guide

kjw142857 committed:
Update user guide

kjw142857 committed:
Update User Guide

kjw142857 committed:
Update User Guide

kjw142857 committed:
Test Github Pages link

kjw142857 committed:
Revert "Test Github Pages link"

This reverts commit 32dcf8ca4b0c38be515fb221a549216354e0b947.

kjw142857 committed:
Update README.md

kjw142857 committed:
Update README.md

kjw142857 committed:
Update user guide

kjw142857 committed:
Increase robustness of error handling

- Fixed a bug where having an (invalid) odd number of lines in the data file would hinder running of the program.
- If data file is corrupted, implemented reloading with an empty file.
